src/rust/src/lib.rs

```diff
@@ -66,8 +73,128 @@ mod rust_bridge {
+//   at N) and execute according to a switch below:
+//
+//   - If E > HI it is an error: core needs an upgrade to proceed.
+//   - If E is in [LO, HI], we run either LO or HI as appropriate.
+//   - if E < LO, we look for exceptions for N:
+//      - If there's an exception for N, we run it
+//      - Otherwise we assume N replays correctly on LO, and run it on LO.
+//
+// - Advancing: when bumping the host crate dependency, the following process
+//   must be followed:
+//
+//   - Assume before the advance that LO=X, HI=Y, and you want to advance the
+//     active window to cover some fresh host version Z.
+//
+//   - If the network is still on HOST_VERSION=X, it must first do an upgrade to
+//     HOST_VERSION=Y before you proceed. You cannot ship a version with HI !=
+//     LO+1 (it is defined that way) since you cannot safely "forget" versions
+//     between LO and HI if they were ever in the field. You must perform this
+//     process for each advance.
+//
+//   - Without loss of generality, then, we can assume the network _has_
+//     upgraded to HOST_VERSION=Y and you want to support Z= Y+1 = X+2. So now
+//     add 1 to `const lo::VERS: usize` below. This redefines LO=Y and HI=Z, and
+//     X is now expired in your local version.
+//
+//   - Update the git revisions for -lo and -hi in the Cargo.toml file, moving
+//     the previous hash for -hi to -lo and adding a new hash for -hi.
+//
+//   - Replay all of history (at least since smart contracts were turned on) to
+//     confirm that there is no drift when expiring X and shifting from
+//     replay-old-ledgers-on-X to replay-on-Y. Since every ledger that will ever
+//     be recorded under HOST_VERSION=X has already been recorded, you just need
+//     to be sure they do not have have semantic drift when replaying on Y.
+//
+//   - If any ledger N replays differently, figure out why. Minimize the
+//     difference as much as possible, then add a new exception-entry to the
+//     multi-version match. Keep doing this until you get perfect replay.
+//
+//   - You are now ready to deploy a version with LO=Y, HI=Z.
+
+impl std::fmt::Display for rust_bridge::BridgeError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+impl std::error::Error for rust_bridge::BridgeError {}
+
+pub(crate) fn invoke_host_function(
+    code_vers: usize,
+    ledger: u32,
+    hf_buf: &rust_bridge::XDRBuf,
+    args_buf: &rust_bridge::XDRBuf,
+    footprint_buf: &rust_bridge::XDRBuf,
+    ledger_entries: &Vec<rust_bridge::XDRBuf>,
+) -> Result<Vec<rust_bridge::Bytes>, Box<dyn std::error::Error>> {
+    if code_vers > hi::VERS {
+        Err(Box::new(rust_bridge::BridgeError::VersionNotYetSupported))
+    } else if code_vers == lo::VERS {
+        lo::contract::invoke_host_function(hf_buf, args_buf, footprint_buf, ledger_entries)
+    } else if code_vers == hi::VERS {
```

Comment:
When will this branch be reached? In other words, who decides the `code_vers` being passed in?
I understand for a historic ledger, it comes from the ledger header. 
For a new transaction, is it always going to be the network-agreed-upon config setting `HOST_VERSION`, which based on my understanding always equal to `lo::VERS`? 

Or is the `hi::VERS` like a "beta" version, where people can choose to run their transaction using it. 

Would it also make sense to always run a new transaction with both versions (maybe only run one of them on chain and another one off-chain to reduce load), so that any drift can be observed right away, rather than right before we want to bump the host crate dependency?